[lld][ELF] Error when deplibs adds new input file after LTO

mysterymath opened the pull request:
Parsing the new input file's symbols might invalidate LTO codegen, but the semantics of deplibs require them to be parsed. Accordingly, report an error unless the file had already been added to the link.

Fixes #56070

mysterymath committed:
[lld][ELF] Error when deplibs adds new input file after LTO

Parsing the new input file's symbols might invalidate LTO codegen, but
the semantics of deplibs require them to be parsed. Accordingly, report
an error unless the file had already been added to the link.

Fixes #56070

mysterymath committed:
Quick update:
 - Consistent plurality in comment
 - errorOrWarn, since an output file may still producible

MaskRay commented on lld/test/ELF/deplibs-lto.s:
```diff
@@ -0,0 +1,39 @@
+# REQUIRES: x86
```
aarch64

MaskRay commented on lld/test/ELF/deplibs-lto.s:
```diff
@@ -0,0 +1,39 @@
+# RUN: llvm-ar rc libfoo.a foo.o
+
+## LTO emits a libcall (`__aarch64_ldadd4_relax`) that is resolved using a
+## library (libdeplibc.a) that contains a `.deplibs` section pointing to a file
```
libdeplibs.a

mysterymath replied:
Done.

MaskRay commented on lld/test/ELF/deplibs-lto.s:
```diff
@@ -0,0 +1,39 @@
+# CHECK: error: input file 'foo.o' added after LTO
+
+## Including the file before LTO prevents the issue.
+
```
Delete the blank line. The convention is to have `##` immediately precede associated RUN lines.

mysterymath replied:
Done.

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -2980,6 +2981,20 @@ template <class ELFT> void LinkerDriver::link(opt::InputArgList &args) {
 
+  // ELF dependent libraries may have introduced new input files after LTO has
+  // completed. This is an error if the files haven't already been parsed, since
+  // it's no longer legal to change the symbol table by parsing them.
```
"no longer legal ..." => changing the symbol table could break the semantic assumption of LTO.

mysterymath replied:
Done.

mysterymath committed:
Address review feedback:

- Require aarch64
- Fix typo
- Remove blank lines in test
- Include reason for illegality in comment

Misc:

- Remove backticks around file and section names
